FxCopAnalyzer fixes for Wox.Core

avneet-kr opened the pull request:
## Summary of the Pull Request

Includes fixes for errors from enabling FxCopAnalyzer.
Note: This is PR # 1 of 3 for Wox.Core, so the change in Wox.Core.csproj to enable FxCopAnalyzer will be included in the 3rd PR

## PR Checklist
* [x] Applies to #5129 
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

Fixes included in this PR:

1. Removed unused function argument from PluginsLoader.cs
2. Check if arguments are null and throw ArgumentNullException
3. Used IsNullOrEmpty to test for empty strings (CA1820: Test for empty strings using string length)
4. FxCop Fixes- added InvarientCulture info, suppressed warnings for general exception types


## Validation Steps Performed

This PR was tested by building the solution and running all unit tests.

avneet-kr committed:
Merge pull request #1 from microsoft/master

Pulling in changes from microsoft/PowerToys

avneet-kr committed:
FxCop Fixes- added InvarientCulture info, suppressed warnings for general exception types

avneet-kr committed:
Used IsNullOrEmpty to test for empty strings (CA1820: Test for empty strings using string length)

avneet-kr committed:
Check if arguments are null and throw ArgumentNullException

avneet-kr committed:
Removed unused function argument from PluginsLoader.cs

crutkas commented:
I see a lot of new ArgumentNullException on methods, is this the correct behavior?

avneet-kr commented:
ArgumentNullExceptions were added to resolve CA1062 (Validate arguments of public methods), according to its recommendation.

somil55 commented on src/modules/launcher/Wox.Core/Plugin/PluginConfig.cs:
```diff
@@ -81,7 +83,9 @@ private static PluginMetadata GetPluginMetadata(string pluginDirectory)
                 // for plugin still use old ActionKeyword
                 metadata.ActionKeyword = metadata.ActionKeywords?[0];
             }
+#pragma warning disable CA1031 // Do not catch general exception types
```
It would be better to use attributes to suppress this error. You can find an example here: https://github.com/microsoft/PowerToys/pull/6958/files#diff-82046de380447e40c4e6c040bdb9643d4f4c019d934ef9c310eba7920891b27aR333

alekhyareddy28 commented on src/modules/launcher/Wox.Core/Plugin/PluginManager.cs:
```diff
@@ -201,7 +216,8 @@ private static List<Result> UpdateResultWithActionKeyword(List<Result> results,
                 {
-                    result.QueryTextDisplay = string.Format("{0} {1}", query.ActionKeyword, result.QueryTextDisplay);
+                    // Using InvariantCulture since this is a command line arg
+                    result.QueryTextDisplay = string.Format(CultureInfo.InvariantCulture, "{0} {1}", query.ActionKeyword, result.QueryTextDisplay);
```
This is a user facing string so shouldn't this be `Currentculture` instead of `InvariantCulture`?

avneet-kr replied:
Thanks @alekhyareddy28. Looking at the available options, would this be `CurrentCulture `or `CurrentUICulture`?

ryanbodrug-microsoft replied:
You should get a compile error if you use the wrong one.

avneet-kr committed:
Addressing comments- Using attributes to suppress errors for general exception types

avneet-kr committed:
Addressing comments- Using attributes to suppress errors

avneet-kr committed:
Addressed comments and changed InvariantCulture to CurrentCulture/Ordinal where appropriate

ryanbodrug-microsoft commented on src/modules/launcher/Wox.Core/Plugin/PluginInstaller.cs:
```diff
@@ -41,15 +41,16 @@ internal static void Install(string path)
-                    .Replace("?", "_")
-                    .Replace("*", "_")
-                    .Replace("|", "_")
+                    .Replace("/", "_", StringComparison.InvariantCulture)
```
StringComparison.Ordinal is a preferable alternative in this case.   Here is some documentation that discusses ordinal vs invariant culture: 

https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings

https://docs.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=netcore-3.1#System_StringComparison_InvariantCulture

ryanbodrug-microsoft commented on src/modules/launcher/Wox.Core/Resource/FontHelper.cs:
```diff
@@ -24,7 +25,9 @@ public static FontWeight GetFontWeightFromInvariantStringOrNormal(string value)
             {
                 return (FontWeight)_fontWeightConverter.ConvertFromInvariantString(value);
             }
+#pragma warning disable CA1031 // Do not catch general exception types
```
We should try to understand what exception we are catching, as this could silently fail due to programmer error.  At the very least we should be logging the exception.

ryanbodrug-microsoft reviewed (changes requested):
Nice work.   I have a couple small suggestions

avneet-kr committed:
Addressing comments - catching and logging specific exception in FontHelper.cs

avneet-kr commented:
Added additional commits to address comments-
- Using attributes to suppress errors for general exception types
- Changed InvariantCulture to CurrentCulture/Ordinal where appropriate
- Caught and logged specific exception in FontHelper.cs

ryanbodrug-microsoft reviewed (approved):
LGTM!
